### DISCUSSION TITLE
How are you meant to strip the `workspace:` protocol when publishing to npm?

### DISCUSSION THREAD

**Discussion by mattfysh** [changesets/changesets#1389] [Q&A]
I can't figure out why `changeset publish` is leaving the `workspace:` prefix in the package.json when publishing to the npm package registry. This leads to npm packages that do not work for anyone on install, e.g.

```
pnpm add @getlang/parser
 ERR_PNPM_WORKSPACE_PKG_NOT_FOUND  In : "@getlang/utils@workspace:0.0.7" is in the dependencies but no package named "@getlang/utils" is present in the workspace
```

I must be misunderstanding something here

**Comment by mattfysh** [comment]
seems like a lot of working around the tools: https://github.com/getlang-dev/get/blob/main/scripts/publish.sh

**Reply by thejasonxie** [reply]
is writing a script the only way to handle symlinks in packages?

**Reply by thejasonxie** [reply]
Anyways, I did the same using bun shell. I don't need to do any git restore since I'm running in this in GitHub actions

```js
// prep-for-publish.ts
// usage: bun ./prep-for-publish.ts

import { $ } from 'bun';

const packages = (await $`ls packages`.text()).split('\n');
if (packages[packages.length - 1].length === 0) packages.pop();

for (const pkg of packages) {
  const pkgJson = JSON.parse(await $`cat packages/${pkg}/package.json`.text());

  console.log(`Preparing ${pkg} for publishing...`);

  // check dependencies
  const deps = Object.keys(pkgJson.dependencies || {});
  for (const dep of deps) {
    if (pkgJson.dependencies[dep].includes('workspace:')) {
      pkgJson.dependencies[dep] = pkgJson.dependencies[dep].replace(
        'workspace:*',
        'latest',
      );
    }
  }

  // check devDependencies
  const devDeps = Object.keys(pkgJson.devDependencies || {});
  for (const devDep of devDeps) {
    if (pkgJson.devDependencies[devDep].includes('workspace:')) {
      pkgJson.devDependencies[devDep] = pkgJson.devDependencies[devDep].replace(
        'workspace:*',
        'latest',
      );
    }
  }

  await $`echo '${JSON.stringify(pkgJson, null, 2)}' > packages/${pkg}/package.json`;
}

```

**Comment by with-heart** [comment]
I'm still in the process of verifying that this is the case but I believe you can use `pnpm -r publish` as a drop-in replacement for `changeset publish` and it will replace `workspace:` references with actual versions when publishing. Maybe @zkochan can provide some insight on the differences?

**Reply by thejasonxie** [reply]
Oh thanks, that's actually a pretty neat pnpm feature. Unfortunately, I'm using bun workspaces and so is mattyfysh but I guess having a pnpm-workspace.yaml in project or generating on in workflow and using `pnpm -r publish` after might work. 

Hopefully bun's `publish` will address this issue when its out.

**Reply by with-heart** [reply]
Oh whoops, I just saw the first line of code block was `pnpm add @getlang/parser` and made the assumption. My bad! Hoping that's the case for `bun` as well.
